### ISSUE TITLE
API for dapp to request network change

### ISSUE THREAD

**Issue by etienneroudeix** [MetaMask/metamask-extension#5101]
Hello,

I would like to connect a metamask user to a custom network directly from my dapp.
The idea is to avoid the user to have to copy the new custom RPC url and paste it in a second page metamask form.

## Context :
Today when a dapp running on a custom Ethereum blockchain wants a user to sign a transaction, the user needs to :
* Copy the custom url
* Open Metamask
* Click the dropdown menu
* Click the parameters button
* Search for the « New RPC URL » form
* Paste the url
* Realize that the other form fields has no connection and are not needed
* Click the save button
* Click to the cross to go back to the main menu

And this a very long process. And metamask does not always remember the custom network.

## My proposal :
The web3 metamask provider accepts a new function to add a custom RPC from the dapp.
The metamask extension shows a window to warn the metamask user that it’s network has been changed. The warning should be enforced if the chainId is 1.
The dapp can ask the user to send a tx to the custom network using `eth_sendTransaction`.

What do you think about this feature ?

**Comment by bdresser** [comment]
@etienneroudeix I think this proposal makes a lot of sense. We've discussed this internally a couple times and would love to collaborate on a solution. 

This could be a good candidate for an EIP, and will likely have some downstream effects on the Custom RPC UI we show in our settings menu. 

See https://github.com/MetaMask/metamask-extension/issues/3604 https://github.com/MetaMask/metamask-extension/pull/5134 https://github.com/MetaMask/metamask-extension/issues/3512

**Comment by danfinlay** [comment]
This is very aligned with my general thinking of MetaMask going forward. For example, it is reminiscent of how we have begun addressing managing a list of tokens:
https://ethereum-magicians.org/t/eip-747-wallet-watchasset/1048/9

Once we merge support for custom chainIds, we could begin implementing this.

Two open questions from that PR:
- Where do we get prices for a custom chain?
- What block explorer do we link to for a custom chain?

Both of these things (prices & explorer links) have previously been neatly trustless parts of MetaMask (you can cross-check what a Dapp claims against them). If Dapps could propose networks, how would we answer these basic questions?

**Comment by eosclassicteam** [comment]
@danfinlay I would like to know the eta schedule for optional chain id support , it has been open from 12 days ago and still no sign of review or merge 😂

**Comment by danfinlay** [comment]
[A good discussion on this topic with a sample spec was raised in January on ethereum-magicians](https://ethereum-magicians.org/t/add-wallet-methods-to-improve-dapp-to-wallet-interaction/1848/31?u=danfinlay)

**Comment by danfinlay** [comment]
New EIP for this: https://github.com/ethereum/EIPs/pull/2015

**Comment by PierreJeanjacquot** [comment]
Hi, the [EIP-2015](https://github.com/ethereum/EIPs/pull/2015) looks really nice for UX.
I need an easy way to ask the user to switch between chains, as described @etienneroudeix the current process is really painful and this EIP will help a lot for the user adoption.
@bdresser Is there any plan at MetaMask to work on this feature?

**Comment by danfinlay** [comment]
Yes, we're looking forward on integrating an API for suggesting network change, but we currently think it might be more appropriate to either/both:
- Integrate it into the initial sign-in prompt (`.enable()`)
- Allow additional network providers to be requested at will (allow multiple network connections without user confirmation).

We are doing work towards both of those right now, and while this might be slightly lower-hanging fruit, these solutions seem more beneficial long-term. Thoughts?

**Comment by etienneroudeix** [comment]
hi @danfinlay 
> Allow additional network providers to be requested at will (allow multiple network connections without user confirmation).

Does it mean a dapp can ask the user to sign a tx on Rinkeby (or any other network)  while his current network is the main network ?
And the opposite ? This last one might be tricky for the user.

**Comment by danfinlay** [comment]
> Does it mean a dapp can ask the user to sign a tx on Rinkeby (or any other network) while his current network is the main network ?
> And the opposite ? This last one might be tricky for the user.

Long term, I think the concept of "connected network" will become less meaningful. What is more important are the assets you hold. Some are test ether, some are real ether. If we represent the asset being transferred correctly, we should be able to provide a good inter-chain dapp development experience without exposing the user to unusual phishing risks.

**Comment by wbt** [comment]
Could we first add a feature allowing the dapp to query the address of the node it is connected to, in addition to being able to get the chainID?  This is much more important for custom RPC addresses.  The user can see what's there, but the dapp currently cannot, and this reduces usability by adding failure paths that are not intuitive for users to debug, even/especially when they can verify that dapp files (e.g. HTML and Javascript) are coming from an expected server.

**Comment by danfinlay** [comment]
Some RPC URLs include private authorization keys as query parameters, so I'm not sure I'd want to leak that information to every site a user visits. I would hope providing chainId/networkId checks as well as the ability to suggest a specific RPC would satisfy most applications' needs. If checking the exact URL of the user's RPC were really necessary, I would be inclined to provide it as an additional sign-in permission, so the user would explicitly grant that information disclosure.

**Comment by wbt** [comment]
How would a dapp know if it needs to request a network change, if it doesn't know what it's on?  
Maybe a feature to get an accurate answer to "Am I connected to RPC URL <x>?" would be a better way to address both.

**Comment by danfinlay** [comment]
For the time being, the `networkId` and `chainId` parameters can be used to identify EVM-based blockchains, and should be used instead of an RPC, since a user may choose to connect to a given blockchain with their own / a custom RPC.

**Comment by Moejoe90** [comment]
any update on implementing EIP-2015?

**Comment by danfinlay** [comment]
We are currently putting more effort into EIP 2255, which will eventually incorporate the parameters from EIP 2015, but allowing the wallet to connect to different networks on different dapps, instead of preserving the fairly broken notion of a single network for all browser tabs.

**Comment by Marksmargon** [comment]
Any estimation when EIP 2255 (which will implement EIP 2015) will implemented? looking forward adding better network switching for our Metamask users.

**Comment by berniegao** [comment]
Hi @danfinlay any ETA for this function to be supported? looking forward to it

**Comment by LucasMoskun** [comment]
I see here that EIP-2255 has been integrated into the browser version 
(https://docs.metamask.io/guide/rpc-api.html#permissions). 

_"Currently, the only permission is eth_accounts, which allows you to access the user's Ethereum address(es). More permissions will be added in the future."_

Is the proposed EIP-2015 functionality still planning to be incorporated as a permission call? If so, is there any idea when this may be?

**Comment by rekmarks** [comment]
Update: We're doing this: https://eips.ethereum.org/EIPS/eip-3085

It's basically like 2015, but without update functionality (for now). We'll have some additional restrictions on the kind of chains that can be added, and will ignore some fields for the time being, but e.g. xDAI, Optimism, and similar EVM-compatible chains with ETH-analogue native currencies (with 18 decimals) will be addable using `wallet_addEthereumChain`.

Basically, fi you can add via the existing custom network RPC form today, you should be able to add via the new RPC method once it ships (soon), so long as the chain ID does not collide with one of the default MetaMask networks.

**Comment by LucasMoskun** [comment]
> Update: We're doing this: https://eips.ethereum.org/EIPS/eip-3085
> 
> It's basically like 2015, but without update functionality (for now). We'll have some additional restrictions on the kind of chains that can be added, and will ignore some fields for the time being, but e.g. xDAI, Optimism, and similar EVM-compatible chains with ETH-analogue native currencies (with 18 decimals) will be addable using `wallet_addEthereumChain`.
> 
> Basically, fi you can add via the existing custom network RPC form today, you should be able to add via the new RPC method once it ships (soon), so long as the chain ID does not collide with one of the default MetaMask networks.

Woooooo! 🍻

**Comment by sameepsi** [comment]
Any updates on this?

**Comment by rekmarks** [comment]
@sameepsi we're close to finishing the feature on our end and should ship soon.

**Comment by sameepsi** [comment]
@rekmarks That's good to hear. Do let us know when this feature is out. We are very much eager to integrate it within [QuickSwap](https://quickswap.exchange/).

**Comment by djsatok** [comment]
@rekmarks , thanks for your work on this issue. **NEAR** is waiting for this too.

**Comment by firstabor** [comment]
@rekmarks wow, that will make everything so much easier especially in those gasintense times right now :)

DEUS is also waiting!

**Comment by Marksmargon** [comment]
**Fuse** is waiting anxiously as well :)
Thanks for pushing this forward @rekmarks

**Comment by djsatok** [comment]
@brad-decker, great work! Congratz!
When do you expect to release it?

**Comment by wbt** [comment]
They merged a PR with the "T12-DONOTMERGE" tag...

**Comment by sameepsi** [comment]
Any updates on this?

**Comment by KaiRo-at** [comment]
> Update: We're doing this: https://eips.ethereum.org/EIPS/eip-3085

Apparently, that doesn't allow to switch users to a built-in network though, which makes operation a lot more difficult for anything that wants to switch and is not using custom chains...

**Comment by masterbd** [comment]
Pancakeswap does this 'network change request', do you guys have any idea how they do it. 
I have to mention that I searched through their github repo, but with no success. Probably some code example will be useful for other new developers.

**Comment by zkipart** [comment]
Hello @masterbd 

You should use wallet_addEthereumChain (https://docs.metamask.io/guide/rpc-api.html#wallet-addethereumchain) for network change request.

Small snippet:

`window.ethereum.request({
            method: 'wallet_addEthereumChain',
            params: [{
                chainId: '0x38',
                chainName: 'Binance Smart Chain',
                nativeCurrency:
                    {
                        name: 'BNB',
                        symbol: 'BNB',
                        decimals: 18
                    },
                rpcUrls: ['https://bsc-dataseed.binance.org/'],
                blockExplorerUrls: ['https://bscscan.com/']
            }]
}`

**Comment by misotsune** [comment]
@zkipart switching between added chains and adding new chains are fundamentally different things though, also metamask doesn't allow to add default chains.

**Comment by KaiRo-at** [comment]
#10597 is the current issue for this request, FWIW - but someone still needs to implement it...
